providers/src/airflow/providers/cncf/kubernetes/callbacks.py

```diff
@@ -83,7 +123,34 @@ def on_pod_completion(*, pod: k8s.V1Pod, client: client_type, mode: str, **kwarg
         pass
 
     @staticmethod
-    def on_pod_cleanup(*, pod: k8s.V1Pod, client: client_type, mode: str, **kwargs):
+    def on_pod_wrapup(
+        *,
+        pod: k8s.V1Pod,
+        client: client_type,
+        mode: str,
+        operator: KubernetesPodOperator,
+        context: Context,
+        **kwargs,
+    ) -> None:
+        """
+        Invoke this callback after all pod completion callbacks but before the pod is deleted.
+
+        :param pod: the completed pod.
+        :param client: the Kubernetes client that can be used in the callback.
+        :param mode: the current execution mode, it's one of (`sync`, `async`).
+        """
+        pass
```

Comment:
The existing code makes a call to find the pod if the callbacks are attached.
https://github.com/apache/airflow/blob/c77c7f003a2458698a1d5a440670b9728783ff78/providers/src/airflow/providers/cncf/kubernetes/operators/pod.py#L614
Honestly I'd prefer if I was sending a stale reference and that it was the responsibility of the callback to get an updated pod if it needs it since we're sending the client too.  Especially since its possible a callback might not implement the `on_pod_completion` method.  So to maintain existing behaviour I'm getting it once.  An alternative might be to test if the `on_pod_completion` is implemented in the callback and get an updated pod for each callback call, but again this assumes we need an updated pod, which might not be the case.

Reply:
As for the need for `on_pod_wrapup` my thought here was that you might have multiple callbacks running before a sidecar container is killed.  Rather than attaching mutliple sidecars to the container, you could have a class that attaches a single sidecar and kills it in the `on_pod_wrapup`, any subclasses of it could pull whatever files they need or run any commands in the pod during the `on_pod_completion` callback.

Reply:
I am not quite sure if I understand you well here. Are you talking about a specific case of a running sidecar?

Reply:
Sorry if I'm not being clear. The way I've been using this so far is that I have one class that extends off the `KubernetesPodOperatorCallback` and does the insertion of a sidecar `on_pod_manifest_created`, killing the sidecar `on_pod_wrapup`, and some code to ensure that the sidecar is only added/killed once.  Then extending off that I have a bunch of other classes that are responsible for doing some actual work with the sidecar, in my case I pull DBT artifacts in the `on_pod_completion` method, then kill the pod in `on_pod_wrapup` before extracting dataset events from DBT artifacts and also uploading them to S3.

Reply:
> An alternative might be to test if the on_pod_completion is implemented in the callback and get an updated pod for each callback call, but again this assumes we need an updated pod, which might not be the case.

True, might be a band-aidy solution

Reply:
Let me see if i understand it correctly, correct me if I am wrong here.

You have a pod running and you want your sidecar to be killed only after the "main" container in your pod completes, is that right?

Reply:
Yep, this is also exactly how XComs works in the kubernetes pod operator.  If you enable xcoms, airflow will insert an xcoms container alongside the base container, and have a shared emptydir volume between them.  When the base container completes airflow will extract the xcoms file from the xcoms container and then send it a kill signal.